crates/apollo-compiler/src/validation/fragment.rs

```diff
@@ -10,12 +16,253 @@ pub fn validate_fragment_definitions(
+                            ),
+                        ))
+                        .help(format!(
+                            "Fragments must be specified on types that exist in the schema. Consider defining `{inline_type_cond}` in the schema."
+                        )),
+                    );
+                }
+            }
+
+            fragment_diagnostics
+        })
+        .collect();
+
+    let operation_inline_fragment_diagnostics = db
+        .operations(file_id)
+        .iter()
+        .flat_map(|operation| {
+            if operation.name().is_some() {
+                let fields = operation.fields(db.upcast());
+                let inline_fragments = fields
+                    .iter()
+                    .flat_map(|field| field.selection_set().inline_fragments());
+
+                inline_fragments.filter_map(|inline_fragment| {
+                    if let Some(inline_type_cond) = inline_fragment.type_condition() {
+                        if !schema_types.contains_key(inline_type_cond) {
+                            Some(
+                                ApolloDiagnostic::new(
+                                    db,
+                                    inline_fragment.loc().into(),
+                                    DiagnosticData::MissingType {
+                                        ty: inline_type_cond.into(),
+                                    },
+                                )
+                                .label(Label::new(
+                                    inline_fragment.loc(),
+                                    format!("`{inline_type_cond}` is defined here but not declared in the schema"),
+                                ))
+                                .help(format!("Fragments must be specified on types that exist in the schema. Consider defining `{inline_type_cond}` in the schema."))
+                            )
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                }).collect::<Vec<_>>()
+            } else {
+                std::iter::empty().collect::<Vec<_>>()
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let mut combined_diagnostics = Vec::with_capacity(
+        diagnostics.len() + operation_inline_fragment_diagnostics.len(),
+    );
+    combined_diagnostics.extend(diagnostics);
+    combined_diagnostics.extend(operation_inline_fragment_diagnostics);
+    combined_diagnostics
+}
+
+pub fn validate_fragment_on_composite_types(
+    db: &dyn ValidationDatabase,
+    file_id: FileId,
+) -> Vec<ApolloDiagnostic> {
+    let mut detected_scalars = std::collections::HashSet::new();
+    let mut detected_inline_scalars = std::collections::HashSet::new();
+    let mut invalid_fragments = std::collections::HashSet::new();
+    let mut invalid_inline_fragments = std::collections::HashSet::new();
```

Comment:
is this the implementation for [this bit of the spec](https://spec.graphql.org/October2021/#sec-Fragments-On-Composite-Types)?

if so, we have a few convenience methods that'd make this a bit easier. Because we always know the `type_condition` on the fragment, we can look up it's type, and then check do an `is_scalar_type_definition()` on it. (might not quite compile)

```rust
let type_cond = def.type_condition();
let ty_def = db.find_type_definition_by_name(type_cond);
let is_scalar = ty_def.is_scalar_type_definition();
```